source/Hyperspace-Model/Retry.class.st

```diff
@@ -0,0 +1,178 @@
+Retry class >> configuredBy: configurationAction [
+
+	^ self new initializeConfiguredBy: configurationAction
+]
+
+{ #category : #accessing }
+Retry class >> defaultMaximumNumberOfAttempts [
+
+	^ 3
+]
+
+{ #category : #evaluating }
+Retry class >> value: aNyladicClosure [
+
+	^ self value: aNyladicClosure configuredBy: [ :retry | retry onException ]
+]
+
+{ #category : #evaluating }
+Retry class >> value: aNiladycBlock configuredBy: configurationAction [
+
+	^ ( self configuredBy: configurationAction ) value: aNiladycBlock
+]
+
+{ #category : #configuring }
+Retry >> backoffExponentiallyWithTimeSlot: aDuration [
+
+	self backoffExponentiallyWithTimeSlot: aDuration using: [ :integer | integer atRandom ]
+]
+
+{ #category : #configuring }
+Retry >> backoffExponentiallyWithTimeSlot: aDuration using: aRandomizingAction [
+
+	"This strategy implements [Exponential Backoff](https://en.wikipedia.org/wiki/Exponential_backoff).
+	
+	`aRandomizingAction` must provide a randomization block that will be evaluated with an integer value 
+	and must answer a random integer between 1 and that value"
+
+	waitActionBetweenRetries := [ :currentAttemptCount | 
+	                            | timeSlotCount |
+	                            "We are applying the randomization function before subtracting one because #atRandom on an integer works in the range [1,self].
+										In other case you will never get a zero time slot"
+	                            timeSlotCount := ( aRandomizingAction value:
+		                                               ( 2 raisedTo: currentAttemptCount ) ) - 1.
+	                            ( aDuration * timeSlotCount ) wait
+	                            ]
+]
+
+{ #category : #configuring }
+Retry >> every: aDuration [
+
+	waitActionBetweenRetries := [ aDuration wait ]
+]
+
+{ #category : #configuring }
+Retry >> ignore: anException [
+
+	exceptionsToIgnore := exceptionsToIgnore , anException
+]
+
+{ #category : #initialization }
+Retry >> initializeConfiguredBy: configurationAction [
+
+	maximumNumberOfAttempts := self class defaultMaximumNumberOfAttempts.
+	waitActionBetweenRetries := [  ].
+	conditionBasedActionBetweenRetries := [  ].
+	exceptionBasedActionBetweenRetries := [  ].
+	exceptionToRetry := ExceptionSet new.
+	exceptionsToIgnore := ExceptionSet new.
+	self when: [ :result | false ].
+	configurationAction value: self
+]
+
+{ #category : #configuring }
+Retry >> on: anException [
+
+	exceptionToRetry := anException
+]
+
+{ #category : #configuring }
+Retry >> on: anException evaluating: aDyadicBlock [
+
+	exceptionToRetry := anException.
+	exceptionBasedActionBetweenRetries := aDyadicBlock
+]
+
+{ #category : #configuring }
+Retry >> onException [
+
+	self on: Exception
+]
+
+{ #category : #configuring }
+Retry >> upTo: aRetryCount [
+
+	AssertionChecker enforce: [ aRetryCount strictlyPositive ]
+		because: 'The maximum number of retries must be strictly positive.'.
+	maximumNumberOfAttempts := aRetryCount + 1
```

Comment:
There's a +1 here and a -1 when using it. Shoudn't they somehow cancel each other out in the code and not be required? Maybe just renaming the variable or something...

Reply:
The variable is used in 4 places.. 2 are decrementing/incrementing and the others not... so it's the same